### PR TITLE
perf(python): Use underlying fileno for Python files when possible

### PR DESCRIPTION
**Pull request by ruihe774**
If a Python file object is passed as the argument of IO functions (e.g. `read_*` & `write_*`), currently it is wrapped into a `PyFileLikeObject`, and each read & write will acquire the GIL and call into Python, which is inefficient. This PR introduces a improvement that extracts the underlying fd through `fileno()` and opens the fd as a Rust `File`, which does not require GIL and have better performance.

### CODE REVIEW

**Comment by ghuls** [comment]
You can't always skip the python read functions as they might do something more complex than just reading the file descriptor as it is: e.g. reading a compressed file format not supported by Polars, or nested compressed files.

```python
$ head -n 1000 test.csv | gzip | gzip | gzip > test.csv.gz.gz.gz


In [19]: with gzip.open("test.csv.gz.gz.gz") as fh:
    ...:     with gzip.open(fh) as fh2:
    ...:         df = pl.read_csv(fh2)
```

**Comment by ruihe774** [comment]
> You can't always skip the python read functions as they might do something more complex than just reading the file descriptor as it is: e.g. reading a compressed file format not supported by Polars, or nested compressed files.

You're right. I limit this optimization to builtin file IO types in 9f4e75c9c541b049f6a515810db0d1b708bba3c5.

File: crates/polars-io/src/mmap.rs [hunk 1]
```diff
@@ -50,14 +49,13 @@ impl Drop for MMapSemaphore {
     }
 }
 
-/// Open a file to get write access. This will check if the file is currently registered as memory mapped.
-pub fn try_create_file(path: &Path) -> PolarsResult<File> {
+/// Check if the file is currently registered as memory mapped.
+pub fn check_not_mapped(path: &Path) -> PolarsResult<()> {
```
Comment (mcrumiller) [inline crates/polars-io/src/mmap.rs:53]: I think `ensure_not_mapped` is a slightly better name for this function, as it's essentially an assert but returns a Result. I would expect a "check" function to return a bool.

**Comment by ruihe774** [comment]
I also add an improvement: use inode number in `MMapSemaphore` instead of path, which better handles hard links and symbolic links.

**Comment by ruihe774** [comment]
> I also add an improvement: use inode number in `MMapSemaphore` instead of path, which better handles hard links and symbolic links.

A better way is to use advisory record locking, and apply it to all file IO, not only mmap, to avoid data race in multithread/multiprocess conditions. I can implement it in the future.

**Review by ritchie46 — APPROVED** [review]
Thanks @ruihe774, this seems like an improvement indeed.

**Pull request merged**